### PR TITLE
Add kx509 client and revamp kx509 service

### PR DESCRIPTION
**Pull request by nicowilliams**
This commit adds support for kx509 in libkrb5, and revamps the KDC's kx509 service (fixing bugs, adding features).

Of note is that kx509 is attempted optimistically by the client, with the certificate and private key stored in the ccache, and optionally in an external PEM or DER file.  This allows users to have PKIX credentials wherever they have Kerberos credentials.  If the service is configured to not require INITIAL tickets, then this functionality amounts to a PKIX credential delegation feature of sorts: wherever a user delegates Kerberos credentials, they can also automatically have PKIX credentials.

A number of related commits are in this PR.  Perhaps those should be separated and pushed first.

There's a lot of work to be done yet, but most if not all of it can be done in later commits.

The full list of commits is:

 - 0f97fec3a hx509: export hx509_parse_private_key (Windows)
 - 4233ef45a hx509: support reading private keys from PEM files
 - adaf680ef hxtool request-create: fix --key argument handling
 - 2fa8e21dd Add more hxtool EKU options, and KeyUsage too
 - a316c624d Add note about racy tests
 - 4d388ed77 Fix hx509_ca_tbs_free() segfault
 - 1091ceaff hxtool: add generate-key command
 - 90dca96c2 hx509: do not crash on missing FILE: name
 - 03a0dbe90 hx509: Add PKCS#8 private key format option
 - 8ba399c0a hx509: create files with safe mode_t
 - 88a01e0e8 hx509: add hx509_certs_destroy()
 - 412283a01 hx509: hx509_context_free(&null_var) should work
 - 64310ba7a hx509: make hx509_get_error_string(NULL, ...) work
 - aaf01a01f hxtool print: fix output device bugs
 - cb10e2e1c Fix missing error checking in lib/krb5/pkinit.c
 - c7b6a7aac kdc: fix invalid free at exit time
 - 2a3937aa9 kdc: fix leaks
 - a232ab702 pkinit: fix memory leak in libkrb5
 - d3f5f0f71 pkinit: fix leak in client
 - 22edd6bb0 klist: fix verbose printing of config entries
 - fac55f869 klist: partial fix of -v --json
 - b8ec34dca Add krb5_cc_configured_default_name()
 - ec1135100 Add kx509 client and revamp kx509 service

(ec1135100 being the HEAD commit).

### CODE REVIEW

**Comment by nicowilliams** [comment]
Missing functionality that needs to be there before pushing:

 - when extracting the private key and certificate from a ccache, update the ccache to list the hx509 store into which they were extracted, and

 - make kdestroy also destroy the hx509 store into which the PKIX creds were stored.

If the PKIX credentials are stored into multiple hx509 stores, we probably want to record each of their names.

**Comment by nicowilliams** [comment]
Windows build fixed.

**Comment by nicowilliams** [comment]
Short-term TBD (after pushing this):

 - add support for using a self-signed CSR or a self-signed Certificate so we can have:
    - proof-of-possession of SPKI
    - algorithm agility
    - a way to express requests for KU/EKU/SANs to be added / left out

Medium-term TBD:

 - finish ECC support in `lib/hx509` and switch to ECDSA as the default

Longer-term TBD:

 - add ed25519 and ed448 support, and switch to that as the default

 - port parts of the UMich CITI `kpkcs11` code to libhx509, with @jaltman's changes for supporting the Windows PKIX API

 - add a PKCS#11 IPC daemon so that we can keep apps from having direct access to private keys

 - similarly, port the Red Hat GSS-API proxy to Heimdal or write similar functionality so we can keep apps from having direct access to credentials

**Comment by lhoward** [comment]
@nicowilliams One could probably shoehorn the PKCS#11 IPC daemon into kcm. Or indeed can you shovel them into the credentials cache so that you can leverage kcm/keyring as-is?

**Comment by nicowilliams** [comment]
@lhoward I'm already setting the cert and priv key in the ccache as cc config entries!  So this should already work with kcm and the linux keyring ccache.

But kcm != isolation.  Clients get to see the session keys, and the tickets (cc config payloads are stored as tickets).

**Comment by lhoward** [comment]
Is the Red Hat GSS proxy a separate thing that could be made to work with Heimdal? I suppose it's not BSD licensed though...

**Comment by nicowilliams** [comment]
@lhoward yeah, it's not GPL, but BSD or MIT, I forget which.  I wrote the XDR for it, and as a result of that cooperation it got licensed with not-GPL.

It's got two components: a daemon, which only uses GSS and so should be trivial to port to Heimdal (or leave it out of tree even), and... invasive libgss mechglue changes that will basically need to be written from scratch for Heimdal.

**Comment by nicowilliams** [comment]
Also, to make this truly useful we'll need to send PRs to OpenSSL and others, or to apps like curl/libcurl, to have some sort of configuration listing `{pkix_cred_store, hostname_pattern_or_domain}` instructing those TLS implementations or TLS apps when to use client certs (for what origins) and which certs.

File: kdc/kx509.c [hunk 1]
```diff
@@ -415,114 +786,147 @@ _kdc_do_kx509(krb5_context context,
...
-
-    ret = kdc_kx509_verify_service_principal(context, cname, sprincipal);
-    if (ret)
+    /* Check that the service name is a valid kx509 service name */
+    if (ret == 0)
+        ret = krb5_ticket_get_server(context, ticket, &sprincipal);
+    if (ret == 0)
+        ret = krb5_unparse_name(context, sprincipal, &sname);
+    if (ret == 0)
+    if (ret == 0)
```
Comment (lhoward) [inline kdc/kx509.c]: superfluous if (ret == 0) here?

File: kuser/klist.c [hunk 2]
```diff
@@ -161,7 +165,28 @@ print_cred_verbose(krb5_context context, krb5_creds *cred, int do_json)
...
+            s = krb5_principal_get_comp_string(context, cred->server, 1);
+            if (strcmp(s, "start_realm") == 0 ||
+                strcmp(s, "anon_pkinit_realm") == 0 ||
+                strcmp(s, "default-ntlm-domain") == 0 ||
+                strcmp(s, "FriendlyName") == 0 ||
+                strcmp(s, "fast_avail") == 0 ||
+                strcmp(s, "kx509store") == 0 ||
+                strcmp(s, "kx509_service_realm") ||
+                strcmp(s, "kx509_service_status") == 0)
+                printf(N_("Configuration item payload: %.*s\n", ""),
```
Comment (lhoward) [inline kuser/klist.c]: maybe these could go in a table or a helper function?

File: doc/standardisation/rfc6717.txt [hunk 3]
```diff
@@ -0,0 +1,339 @@
+
+
+
+
+
+
+Internet Engineering Task Force (IETF)                       K. Zeilenga
+Request for Comments: 6171                                 Isode Limited
```
Comment (lhoward) [inline doc/standardisation/rfc6717.txt]: This is RFC6171 not 6717

Comment (nicowilliams) [reply doc/standardisation/rfc6717.txt]: Hah, whoops, thanks!

File: kuser/klist.c (hunk 2 above)
Comment (nicowilliams) [reply kuser/klist.c]: In klist or in the library?

**Comment by nicowilliams** [comment]
This was reviewed off-github by @vdukhovni.

**Pull request merged**